src/lib/gssapi/spnego/negoex_util.c

```diff
@@ -0,0 +1,810 @@
+                   data1, data2, data3, guid[8], guid[9], guid[10], guid[11],
+                   guid[12], guid[13], guid[14], guid[15]);
+}
+
+static char *
+guid_to_string(const uint8_t guid[GUID_LENGTH])
+{
+    struct k5buf buf;
+
+    k5_buf_init_dynamic(&buf);
+    add_guid(&buf, guid);
+    return buf.data;
+}
+
+/* Check that the described vector lies within the message, and return a
+ * pointer to its first element. */
+static inline const uint8_t *
+vector_base(size_t offset, size_t count, size_t width,
+            const uint8_t *msg_base, size_t msg_len)
+{
+    if (offset > msg_len || count > (msg_len - offset) / width)
+        return NULL;
+    return msg_base + offset;
+}
+
+/* Trace a received message.  Call after the context sequence number is
+ * incremented. */
+static void
+trace_received_message(spnego_gss_ctx_id_t ctx,
+                       const struct negoex_message *msg)
+{
+    struct k5buf buf;
+    uint16_t i;
+    char *info = NULL;
+
+    if (msg->type == INITIATOR_NEGO || msg->type == ACCEPTOR_NEGO) {
+        k5_buf_init_dynamic(&buf);
+        for (i = 0; i < msg->u.n.nschemes; i++) {
+            add_guid(&buf, msg->u.n.schemes + i * GUID_LENGTH);
+            if (i + 1 < msg->u.n.nschemes)
+                k5_buf_add(&buf, " ");
+        }
+        info = buf.data;
+    } else if (msg->type == INITIATOR_META_DATA ||
+               msg->type == ACCEPTOR_META_DATA ||
+               msg->type == CHALLENGE || msg->type == AP_REQUEST) {
+        info = guid_to_string(msg->u.e.scheme);
+    } else if (msg->type == VERIFY) {
+        info = guid_to_string(msg->u.v.scheme);
+    } else if (msg->type == ALERT) {
+        info = guid_to_string(msg->u.a.scheme);
+    }
+
+    if (info == NULL)
+        return;
+
+    TRACE_NEGOEX_INCOMING(ctx->kctx, ctx->negoex_seqnum - 1,
+                          typestr(msg->type), info);
+    free(info);
+}
+
+/* Trace an outgoing message with a GUID info string.  Call after the context
+ * sequence number is incremented. */
+static void
+trace_outgoing_message(spnego_gss_ctx_id_t ctx, enum message_type type,
+                       const uint8_t guid[GUID_LENGTH])
+{
+    char *info = guid_to_string(guid);
+
+    if (info == NULL)
+        return;
+    TRACE_NEGOEX_OUTGOING(ctx->kctx, ctx->negoex_seqnum - 1, typestr(type),
+                          info);
+    free(info);
+}
+
+static OM_uint32
+parse_nego_message(OM_uint32 *minor, struct k5input *in,
+                   const uint8_t *msg_base, size_t msg_len,
+                   struct nego_message *msg)
+{
+    const uint8_t *p;
+    uint64_t protocol_version;
+    uint32_t extension_type;
+    size_t offset, count, i;
+
+    p = k5_input_get_bytes(in, sizeof(msg->random));
+    if (p != NULL)
+        memcpy(msg->random, p, sizeof(msg->random));
+    protocol_version = k5_input_get_uint64_le(in);
+    if (protocol_version != 0) {
+        *minor = ERR_NEGOEX_UNSUPPORTED_CRITICAL_EXTENSION; /* XXX */
```

Comment:
Comment should probably be more descriptive.

Reply:
I think Luke might have left that marker as a reminder to allocate a new error.  I can do that.

There are possibly too many different errors for "a protocol constraint was violated".  But a separate error for "the other side used a major version I don't recognize" probably makes sense.

Reply:
Yes, that would've been a note to self to allocate a new error.